dateparser/date.py

```diff
@@ -87,55 +90,145 @@ def get_last_day_of_month(year, month):
+
+            if not ('%y' in date_format or '%Y' in date_format):
+                today = datetime.today()
+                date_obj = date_obj.replace(year=today.year)
+
+            return {'date_obj': date_obj, 'period': period}
     else:
-        if final_call:
-            return data
+        return {'date_obj': None, 'period': period}
+
+
+class _DateLanguageParser(object):
+    def __init__(self, language, date_string, date_formats):
+        self.language = language
+        self.date_string = date_string
+        self.date_formats = date_formats
+        self._translated_date = None
+        self._translated_date_with_formatting = None
+
+    @classmethod
+    def parse(cls, language, date_string, date_formats=None):
+        instance = cls(language, date_string, date_formats)
+        return instance._parse()
+
+    def _parse(self):
+        for parser in (
+            self._try_timestamp,
+            self._try_freshness_parser,
+            self._try_given_formats,
+            self._try_dateutil_parser,
+            self._try_hardcoded_formats,
+        ):
+            date_obj = parser()
+            if self._is_valid_date_obj(date_obj):
+                return date_obj
+        else:
+            return None
+
+    def _try_timestamp(self):
+        return {
+            'date_obj': get_date_from_timestamp(self.date_string),
+            'period': 'day',
+        }
+
+    def _try_freshness_parser(self):
+        return freshness_date_parser.get_date_data(self._get_translated_date())
+
+    def _try_dateutil_parser(self):
+        try:
+            date_obj = date_parser.parse(self._get_translated_date())
+            return {
+                'date_obj': date_obj.replace(tzinfo=None),
+                'period': 'day',
+            }
+        except ValueError:
+            return None
+
+    def _try_given_formats(self):
+        if not self.date_formats:
+            return
+
+        return parse_with_formats(self._get_translated_date_with_formatting(), self.date_formats)
+
+    def _try_hardcoded_formats(self):
+        hardcoded_date_formats = [
+            '%B %d, %Y, %I:%M:%S %p',
+            '%b %d, %Y at %I:%M %p',
+            '%d %B %Y %H:%M:%S',
+            '%A, %B %d, %Y',
+        ]
+        try:
+            return parse_with_formats(self._get_translated_date_with_formatting(), hardcoded_date_formats)
+        except TypeError:
+            return None
+
+    def _get_translated_date(self):
+        if self._translated_date is None:
+            self._translated_date = self.language.translate(self.date_string, keep_formatting=False)
+        return self._translated_date
+
+    def _get_translated_date_with_formatting(self):
+        if self._translated_date_with_formatting is None:
+            self._translated_date_with_formatting = self.language.translate(self.date_string, keep_formatting=True)
+        return self._translated_date_with_formatting
+
+    def _is_valid_date_obj(self, date_obj):
```

Comment:
A agree, your code is more pythonic. Thanks!
Although it would require adding `sorted(date_obj.keys()) == ['date_obj', 'period']` to do the same job.